keps/sig-node/3063-dynamic-resource-allocation/README.md

```diff
@@ -0,0 +1,2164 @@
+
+- A resource is no longer needed when it has a `DeletionTimestamp`. It must not
+  be freed yet when it is still in use.
+
+Some of the race conditions that need to be handled are:
+
+- A ResourceClaim gets created and deleted again while the resource driver
+  starts allocating it. Before it actually starts doing anything, the
+  driver adds a finalizer. Either adding the finalizer or removing the
+  ResourceClaim win. If the driver wins, it continues with allocation
+  and can either complete or abort the operation when it notices the non-nil
+  DeletionTimestamp. Otherwise, allocation gets aborted immediately.
+
+  What this avoids is the situation where an allocation succeed without having
+  an object where the result can be stored. The driver can also be killed at
+  any time: when it restarts, the finalizer indicates that allocation may be in
+  progress and has to be completed or aborted.
+
+- In a cluster with multiple scheduler instances, two pods might get
+  scheduled concurrently by different schedulers. When they reference
+  the same ResourceClaim which may only get used by one pod at a time,
+  only one pod can be scheduled.
+
+  Both schedulers try to add their pod to the `ReservedFor` field, but only the
+  update that reaches the API server first gets stored. The other one fails
+  with a conflict error and the scheduler which issued it knows that it must
+  put the pod back into the queue, waiting for the ResourceClaim to become
+  usable again.
+
+### Custom parameters
+
+To support arbitrarily complex parameters, both ResourceClass and ResourceClaim
+contain one field of type
+[v1.TypedLocalObjectReference](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.23/#typedlocalobjectreference-v1-core)
+which references a separate object. This information is sufficient for generic clients to
+retrieve the parameters. For ResourceClass, that object must be
+cluster-scoped. For ResourceClaim, it must be in the same namespace as the
+ResourceClaim and thus the Pod. Which objects a resource driver accepts as parameters depends on
+the driver.
+
+This approach was chosen because then validation of the parameters can be done
+with a CRD and that validation will work regardless of where the parameters
+are needed.
+
+A resource driver may support modification of the parameters while a resource
+is in use ("online resizing"). It may update the ResourceClaim status to
+reflect the modified state, for example by increasing the number of concurrent
+users. However, the state must not be modified such that a user of the resource
+no longer has access.
+
+Parameters may get deleted before the ResourceClaim or ResourceClass that
+references them. In that case, a pending resource cannot be allocated until the
+parameters get recreated. An allocated resource must remain usable and freeing
+it must be possible. To support this, resource drivers must copy all relevant
+information into the ResourceClaim status when allocating it.
```

Comment:
Updated:

An allocated resource must remain usable and
deallocating it must be possible. To support this, resource drivers must copy
all relevant information:
- For usage, the `AllocationResult.Attributes` can be hold the copied information
  because the ResourceClaim and thus this field must exist.
- For deallocation, drivers should use some other location to handle
  cases where a user force-deletes a ResourceClaim or the entire
  cluster gets removed.